browser-switch/src/main/java/com/braintreepayments/browserswitch/BrowserSwitchClient.java

```diff
@@ -0,0 +1,332 @@
+        String errorMessage = assertCanPerformBrowserSwitch(requestCode, appContext, intent);
+        if (errorMessage == null) {
+            BrowserSwitchRequest request = new BrowserSwitchRequest(
+                    requestCode, intent.getData(), BrowserSwitchRequest.PENDING);
+            persistentStore.putActiveRequest(request, appContext);
+            appContext.startActivity(intent);
+        } else {
+            BrowserSwitchResult result =
+                new BrowserSwitchResult(BrowserSwitchResult.STATUS_ERROR, errorMessage);
+            listener.onBrowserSwitchResult(requestCode, result, null);
+        }
+    }
+
+    private String assertCanPerformBrowserSwitch(int requestCode, Context context, Intent intent) {
+        String errorMessage = null;
+
+        if (!isValidRequestCode(requestCode)) {
+            errorMessage = "Request code cannot be Integer.MIN_VALUE";
+        } else if (!isConfiguredForBrowserSwitch(context)) {
+            errorMessage =
+                "The return url scheme was not set up, incorrectly set up, " +
+                "or more than one Activity on this device defines the same url " +
+                "scheme in it's Android Manifest. See " +
+                "https://github.com/braintree/browser-switch-android for more " +
+                "information on setting up a return url scheme.";
+        } else if (!canOpenUrl(context, intent)) {
+            StringBuilder messageBuilder = new StringBuilder("No installed activities can open this URL");
+            Uri uri = intent.getData();
+            if (uri != null) {
+                messageBuilder.append(String.format(": %s", uri.toString()));
+            }
+            errorMessage = messageBuilder.toString();
+        }
+
+        return errorMessage;
+    }
+
+    private boolean isValidRequestCode(int requestCode) {
+        return requestCode != Integer.MIN_VALUE;
+    }
+
+    private boolean isConfiguredForBrowserSwitch(Context context) {
+        Intent browserSwitchActivityIntent =
+            config.createIntentForBrowserSwitchActivityQuery(returnUrlScheme);
+        return activityFinder.canResolveActivityForIntent(context, browserSwitchActivityIntent);
+    }
+
+    private boolean canOpenUrl(Context context, Intent intent) {
+        return activityFinder.canResolveActivityForIntent(context, intent);
+    }
+
+    /**
+     * Deliver a pending browser switch result to an Android fragment that is also a BrowserSwitchListener.
+     * We recommend you call this method in onResume to receive a browser switch result once your
+     * app has re-entered the foreground.
+     *
+     * Cancel and Success results will be delivered only once. If there are no pending
+     * browser switch results, this method does nothing.
+     *
+     * @param fragment the BrowserSwitchListener that will receive a pending browser switch result
+     */
+    void deliverResult(Fragment fragment) {
```

Comment:
Should these `deliverResult()`s all be marked as public?

Reply:
We can. I was marking methods public on an as-needed basis.